src/server/dev_checks.ts

```diff
@@ -0,0 +1,255 @@
+        message:
+          `Route at ${route.name} must have a handler or component. It's possible you're missing a default export.`,
+        link: route.url,
+      }];
+    }
+    return [];
+  });
+}
+
+/** Asserts that the usage of the `staticDir` option is safe. */
+export function assertStaticDirSafety(
+  dir: string,
+  defaultDir: string,
+): CheckResult[] {
+  const results: CheckResult[] = [];
+
+  if (dir === defaultDir) {
+    results.push({
+      category: CheckCategory.StaticDirectory,
+      message:
+        `You cannot use the default static directory as a static override. Please choose a different directory.`,
+    });
+  }
+
+  if (dir && existsSync(defaultDir)) {
+    results.push({
+      category: CheckCategory.StaticDirectory,
+      message:
+        `You cannot have both a static override and a default static directory. Please remove the default static directory.`,
+      link: defaultDir,
+    });
+  }
+
+  return results;
+}
+
+/** Asserts that here are no route conflicts with static files. */
+export function assertNoStaticRouteConflicts(
+  routes: Route[],
+  staticFiles: StaticFile[],
+): CheckResult[] {
+  const routePatterns = new Set(routes.map((route) => route.pattern));
+
+  return staticFiles.flatMap((staticFile) => {
+    if (routePatterns.has(staticFile.path)) {
+      return [{
+        category: CheckCategory.StaticFileConflict,
+        message:
+          `Static file conflict: A file exists at ${staticFile.path} which matches a route pattern. Please rename the file or change the route pattern.`,
+        link: staticFile.localUrl.pathname,
+      }];
+    }
+    return [];
+  });
+}
+
+/** Asserts that each plugin, which uses `plugin.render`, calls `ctx.render()` once. */
+export function assertPluginsCallRender(plugins: Plugin[]): CheckResult[] {
+  const results: CheckResult[] = [];
+
+  plugins.forEach((plugin) => {
+    if (typeof plugin.render === "function") {
+      const renderSpy = spy(() => ({
+        htmlText: "",
+        requiresHydration: false,
+      }));
+      plugin.render({ render: renderSpy });
+      try {
+        assertSpyCalls(renderSpy, 1);
+      } catch {
+        results.push({
+          category: CheckCategory.PluginRender,
+          message: `Plugin ${plugin.name} must call ctx.render() exactly once.`,
+        });
+      }
+    }
+  });
+
+  return results;
+}
+
+/** Asserts that each plugin, which uses `plugin.renderAsync`, calls `ctx.renderAsync()` once. */
+export function assertPluginsCallRenderAsync(
+  plugins: Plugin[],
+): CheckResult[] {
+  const results: CheckResult[] = [];
+
+  for (const plugin of plugins) {
+    if (typeof plugin.renderAsync === "function") {
+      try {
+        const renderAsyncSpy = spy(() => {
+          return Promise.resolve({
+            htmlText: "",
+            requiresHydration: false,
+          });
+        });
+
+        plugin.renderAsync({ renderAsync: renderAsyncSpy }).then(() => {});
```

Comment:
Nit: the `.then(() => {});` doesn't seem to do anything

Reply:
added a comment for the "why", as i wanted to keep each of the checks as sync operations.
however whenever i tried to convert it to something like this, i ran into timing issues, where the check resolved before the plugin was done:  
```ts
plugin.renderAsync({ renderAsync: renderAsyncSpy }).then(() => {
  assertSpyCalls(renderAsyncSpy, 1)
}).catch(() => {
  results.push({
    category: CheckCategory.PluginRenderAsync,
    message:
      `Plugin ${plugin.name} must call ctx.renderAsync() exactly once.`,
  });
})
```